internal/overlord/servstate/manager.go

```diff
@@ -39,34 +38,97 @@ func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string) (*Se
 	return manager, nil
 }
 
-func (m *ServiceManager) reloadSetup() error {
-	setup, err := setup.ReadDir(m.pebbleDir)
+func (m *ServiceManager) reloadPlan() error {
+	p, err := plan.ReadDir(m.pebbleDir)
 	if err != nil {
 		return err
 	}
-	flattened, err := setup.Flatten()
+	m.plan = p
+	return nil
+}
+
+// Plan returns the configuration plan.
+func (m *ServiceManager) Plan() (*plan.Plan, error) {
+	releasePlan, err := m.acquirePlan()
 	if err != nil {
-		return err
+		return nil, err
 	}
-	m.setup = setup
-	m.flattened = flattened
-	return nil
+	defer releasePlan()
+	return m.plan, nil // TODO: copy?
+}
+
+// MergeLayer merges the given layer YAML into the dynamic layers, returning
+// the new layer's "order" (won't increase if adding another dynamic layer).
+func (m *ServiceManager) MergeLayer(layerYAML []byte) (int, error) {
```

Comment:
This probably needs terminology tuning depending on previous points. The implementation below looks great, though.

Reply:
Fixed: `CombineLayer`.